internal/owner: introduce OWNER mechanism in codebase

otan opened the pull request:
Deprecated; see https://github.com/cockroachdb/cockroach/pull/52053.

----

This commit introduces a concept of owner in the code base, in the
form of an OWNERS file in relevant subdirectories, which feeds off
a list of teams in TEAMS. This will give us new capabilities to
effectively triage items. The TEAMS will later be populated with
useful information such as Github board ID, contact teams, etc.
These are all in YAML format.

Also introduce a `whoownsit` command, which takes in arguments and
produces the owner for each team.

I've populated a few placeholder teams for now -- we can more more
specific about it later.

See documentation for more details.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51880)
<!-- Reviewable:end -->

petermattis reviewed (commented):
I'm thumbs-up on the general idea of documenting ownership in the code via an OWNERS mechanism. 

Two bikeshed questions:

* Is YAML the right format? I feel like this might be a meme where the answer is no. I'd note that `.github/CODEOWNERS` uses the `.gitignore` syntax. 
* Even if we stick with YAML, do we need the `.yaml` suffix? This might just be my historical context from Google, but I'm used to seeing `OWNERS`, not `OWNERS.yaml`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51880#-:-MDFeIidE8PzjwYCoPRw:bg8qk76)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
* Is YAML the right format? I feel like this might be a meme where the answer is no. I'd note that .github/CODEOWNERS uses the .gitignore syntax.

hah, we can choose. we're not using the Github tooling for this (it's restrictive in that only *last* match counts), so I opted for something more... flexible.

* Even if we stick with YAML, do we need the .yaml suffix? This might just be my historical context from Google, but I'm used to seeing OWNERS, not OWNERS.yaml.

nope, i can remove the prefix.

tbg commented on pkg/OWNERS:
```diff
@@ -0,0 +1,4 @@
+files:
+  Makefile:
+    owners:
```
Do we need this extra level? Can't we just hang the teams right off the pattern?

otan replied:
I was thinking we could potentially add more attributes, e.g.

```
files:
   Makefile:
      owners:
       - security
      always_review: true # do not allow PRs to merge unless it has been reviewed by security, but flexible on this
```

otan replied:
if we think that's too complicated though, happy to take it out

tbg commented on pkg/geo/OWNERS:
```diff
@@ -0,0 +1,2 @@
+owners:
+  - geospatial
```
ditto here. Is there some `yaml` restriction that we can't have this be just

```
geospatial
```

I think that should work as this is a valid yaml document (representing `"geospatial"`).

With two teams,

```
- geospatial
- lololo
```

I dislike that this can't be

```
geospatial
lololo
```

as this is what a non-yaml'ers intuition would dictate.

otan replied:
That's very complicated to parse, especially with multiple file owners :\ (we unfortuantely can't get away from this because the sql package will require fine grained permissions)

tbg commented:
> hah, we can choose. we're not using the Github tooling for this (it's restrictive in that only last match counts), so I opted for something more... flexible.

Could you explain this with an example? From my (limited) experience, yaml is not very intuitive (ten ways to write the same thing, maps vs lists needs some getting used to, etc), and it allows you to express crazy stuff (not that we'd do that here, but still makes me uneasy), so emotionally I like to stay away from it for uses where we expect $everyone to make changes seamlessly.

I looked briefly into `.gitignore` [syntax](https://git-scm.com/docs/gitignore). There are ways to write `.gitignore` files that no human can understand (use negations, etc) but we could limit to a subset. But there's no obvious way to express ownership this way; all we do is match files and I doubt we want to go to a one-file-per-owner pattern.

I think sticking with simple yaml is the way to go for now (or something else that lets you express lists and maps without typing too much), but I think we should do some extra work to make sure the "standard" owner files look minimal and intuitive (i.e. don't need to know yaml most of the time, and in the files where you do, there's typically enough of it already to get by).

Thanks for working on this @otan.

tbg commented on pkg/OWNERS:
```diff
@@ -0,0 +1,4 @@
+files:
```
You intentionally decided against a monolith file, right? We could have something at the repo root that has lines like

```
**/kv: kv
```

tbg replied:
PS the above pattern is one that I would use -- anything that has `kv` in it is owned by `kv`. (Or perhaps even: anything in a package starting with `kv` is). Curious what this would look like properly expressed. Will there be terrible regexps? I'd much rather deal with some of these `**` pattern matchers that were made for file matching.

otan replied:
I chose the "spread out" model because one file can get way too big. not being able to fine grain by package without parsing a monolithic file can be quite gnarly for a human. keeping ownership as close to the package as possible is my preference.

otan replied:
That said, since pkg/kv/OWNERS has

```
owners:
- kv
```

in it, which means that and its subdirectories are all KV.

otan replied:
(thereoetically in this model, `**/kv` works in `pkg/OWNERS`, but again trying to avoid the monolithic file.

otan committed:
internal/owner: introduce OWNER mechanism in codebase

This commit introduces a concept of owner in the code base, in the
form of an OWNERS file in relevant subdirectories, which feeds off
a list of teams in TEAMS. This will give us new capabilities to
effectively triage items. The TEAMS will later be populated with
useful information such as Github board ID, contact teams, etc.
These are all in YAML format.

Also introduce a `whoownsit` command, which takes in arguments and
produces the owner for each team.

I've populated a few placeholder teams for now -- we can more more
specific about it later.

See documentation for more details.

Release note: None

otan commented:
based on feedback so far, people seem to prefer
* the .gitignore syntax
* a monolithic file for ownership

whilst i'm not totally sold, it makes for easier coding and less "work" so i'm going to close this for now. i'll try re-use some of it for `whoownsit` / a `TEAMS.yaml` mapping of config later.

otan commented:
i've replaced this logic with https://github.com/cockroachdb/cockroach/pull/52053.
